### ISSUE TITLE
Implement `cpu_serial_kernel` and `cpu_serial_kernel_vec` for TensorIterator

### ISSUE THREAD

**Issue by VitalyFedyunin** [pytorch/pytorch#24472]
There are ~20 instances of CPU_tensor_apply* use. They need to be converted to `TensorIterator`. We cannot do it without serial loops support.

Ideally we also need `gpu_serial` loops. But it is low-pri.

**Comment by fmassa** [comment]
Do we want to create a new version of the loop, or instead pass an option to TensorIterator during build to force serial path?
This way, we could potentially reuse the same cpu_kernel and cpu_kernel_vec?

Or do we need a completely different approach for serial implementation?

**Comment by VitalyFedyunin** [comment]
Implementing new version of the loop looks better in terms of potential performance impact, especially because TensorIterator doesn't care about traverse order and parallelism, it is loop abstraction that does it

**Comment by fmassa** [comment]
Can you expand a bit about traverse order and parallelism?

From my understanding, cpu_kernel calls for_each
https://github.com/pytorch/pytorch/blob/76716f6c0644f66e5984ab6686ca19f3702039a8/aten/src/ATen/native/cpu/Loops.h#L170
with a basic loop function
https://github.com/pytorch/pytorch/blob/76716f6c0644f66e5984ab6686ca19f3702039a8/aten/src/ATen/native/cpu/Loops.h#L172
Which is sequentially iterating over the elements of the array
https://github.com/pytorch/pytorch/blob/76716f6c0644f66e5984ab6686ca19f3702039a8/aten/src/ATen/native/cpu/Loops.h#L99
Which in turn are called from 0 to numel
https://github.com/pytorch/pytorch/blob/50161f3b3c95635c82d43add453a7c1692eface1/aten/src/ATen/native/TensorIterator.cpp#L414

This takes the ordering after stride permutation into account though.
Is that what you meant by traverse order?

**Comment by VitalyFedyunin** [comment]
`for_each` currently calls `at::parallel_for`
 https://github.com/pytorch/pytorch/blob/50161f3b3c95635c82d43add453a7c1692eface1/aten/src/ATen/native/TensorIterator.cpp#L416

We need another loop which will call `serial_for_each` as `CPU_tensor_apply*` ops are supposed to run in a single thread (we had  `CPU_tensor_parallel_apply*` for muti-threaded cases, and it is already removed in favor of `cpu_loop`)

**Comment by fmassa** [comment]
@VitalyFedyunin `for_each` also calls into the serial version. What I propose is to add an extra bool flag to TensorIterator, that forces dispatch to the serial code path, instead of having to add a new loop (which cpu_kernel_serial implies)

**Comment by VitalyFedyunin** [comment]
> @VitalyFedyunin `for_each` also calls into the serial version. What I propose is to add an extra bool flag to TensorIterator, that forces dispatch to the serial code path, instead of having to add a new loop (which cpu_kernel_serial implies)
> 
> I might be missing something in here though, but I'd love to understand what

This is also acceptable option, but it will require checking if compiler can optimize out this unnecessary field checks.

**Comment by fmassa** [comment]
I thought one extra bool check in https://github.com/pytorch/pytorch/blob/50161f3b3c95635c82d43add453a7c1692eface1/aten/src/ATen/native/TensorIterator.cpp#L413 wouldn't be too bad, given that it would happen only once for the whole kernel?

**Comment by VitalyFedyunin** [comment]
I tend to agree, but still need to check how much impactful for https://github.com/pytorch/pytorch/blob/200cb836f1e31c877e9ac165c326927fa42a28fa/aten/src/ATen/native/TensorIteratorReduce.cpp#L16 this change might be
